akka-actor/src/main/scala/akka/util/LineNumbers.scala

```diff
@@ -0,0 +1,264 @@
+package akka.util
+
+import java.io.DataInputStream
+import java.io.BufferedInputStream
+import scala.annotation.tailrec
+
+object LineNumbers {
+  sealed trait Result
+  case object NoSourceInfo extends Result
+  case class SourceFile(filename: String) extends Result {
+    override def toString = filename
+  }
+  case class SourceFileLines(filename: String, from: Int, to: Int) extends Result {
+    override def toString = s"$filename:$from-$to"
+  }
+
+  private class Constants(val fwd: Map[Int, String], val rev: Map[String, Int]) {
+    def this() = this(Map.empty, Map.empty)
+    def apply(idx: Int): String = fwd(idx)
+    def apply(str: String): Int = rev(str)
+    def updated(idx: Int, str: String): Constants =
+      if (rev contains str) new Constants(fwd.updated(idx, str), rev)
+      else new Constants(fwd.updated(idx, str), rev.updated(str, idx))
+    def contains(str: String): Boolean = rev contains str
+  }
+
+  final val debug = false
+}
+
+class LineNumbers {
+  import LineNumbers._
+
+  // FIXME: this needs memoization with an LRU cache
```

Comment:
I quickly skimmed it and it is a lot of code; I’ll leave this fixme here and settle on an implementation later, thanks for the pointers in any case!